File: ppcbudget/aggregate.py
"""Roll campaign-days up across multiple exports to find chronic offenders.

A campaign out of budget 8 hours a day for two weeks is a bigger problem than
one that spiked to 23 hours once, so `chronic_score` weights recurrence and
streak length alongside average severity.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from statistics import median

from .scoring import CampaignDay

OOB_DAY_THRESHOLD_MIN = 60  # a day "counts" once an hour is lost
STREAK_CEILING = 7

W_RECURRENCE, W_MEAN, W_STREAK = 0.40, 0.35, 0.25


@dataclass(slots=True)
class DayPoint:
    date_key: str
    oob_hours: float
    in_hours: float
    paused_hours: float
    episodes: int
    severity: float
    first_oob_min: int | None


@dataclass(slots=True)
class CampaignRollup:
    campaign: str
    days_observed: int = 0
    days_with_oob: int = 0
    total_oob_hours: float = 0.0
    mean_oob_hours: float = 0.0
    mean_in_hours: float = 0.0  # hours per day the campaign could actually spend
    mean_paused_hours: float = 0.0
    median_oob_hours: float = 0.0
    max_oob_hours: float = 0.0
    total_episodes: int = 0
    mean_first_oob_min: float | None = None
    recurrence_rate: float = 0.0
    streak_current: int = 0
    streak_max: int = 0
    trend_slope: float = 0.0
    chronic_score: float = 0.0
    mean_severity: float = 0.0
    dominant_diagnosis: str = ""
    worst_date: str = ""
    total_lost_spend: float | None = None
    total_lost_sales: float | None = None
    per_day: list[DayPoint] = field(default_factory=list)

    @property
    def trend_label(self) -> str:
        if abs(self.trend_slope) < 0.05:
            return "flat"
        return "worsening" if self.trend_slope > 0 else "improving"


def _ols_slope(values: list[float]) -> float:
    """Least-squares slope of y against its index. Zero for fewer than 2 points."""
    n = len(values)
    if n < 2:
        return 0.0
    mean_x = (n - 1) / 2
    mean_y = sum(values) / n
    denom = sum((i - mean_x) ** 2 for i in range(n))
    if denom == 0:
        return 0.0
    return sum((i - mean_x) * (v - mean_y) for i, v in enumerate(values)) / denom


def rollup(days: list[CampaignDay]) -> list[CampaignRollup]:
    by_campaign: dict[str, list[CampaignDay]] = {}
    for d in days:
        by_campaign.setdefault(d.campaign, []).append(d)

    out: list[CampaignRollup] = []
    for campaign, entries in by_campaign.items():
        entries.sort(key=lambda d: d.date_key)
        hours = [d.oob_hours for d in entries]
        firsts = [d.first_oob_min for d in entries if d.first_oob_min is not None]

        r = CampaignRollup(campaign=campaign, days_observed=len(entries))
        r.per_day = [
            DayPoint(d.date_key, d.oob_hours, d.in_hours, d.paused_hours,
                     d.episodes_merged, d.severity, d.first_oob_min)
            for d in entries
        ]
        r.days_with_oob = sum(1 for d in entries if d.oob_min >= OOB_DAY_THRESHOLD_MIN)
        r.total_oob_hours = sum(hours)
        r.mean_oob_hours = r.total_oob_hours / len(entries)
        r.mean_in_hours = sum(d.in_hours for d in entries) / len(entries)
        r.mean_paused_hours = sum(d.paused_hours for d in entries) / len(entries)
        r.median_oob_hours = median(hours)
        r.max_oob_hours = max(hours)
        r.total_episodes = sum(d.episodes_merged for d in entries)
        r.mean_first_oob_min = (sum(firsts) / len(firsts)) if firsts else None
        r.recurrence_rate = r.days_with_oob / len(entries)

        streak = 0
        for d in entries:
            if d.oob_min >= OOB_DAY_THRESHOLD_MIN:
                streak += 1
                r.streak_max = max(r.streak_max, streak)
            else:
                streak = 0
        r.streak_current = streak
        r.trend_slope = _ols_slope(hours)

        r.chronic_score = 100 * (
            W_RECURRENCE * r.recurrence_rate
            + W_MEAN * min(r.mean_oob_hours / 24, 1.0)
            + W_STREAK * min(r.streak_max, STREAK_CEILING) / STREAK_CEILING
        )

        r.mean_severity = sum(d.severity for d in entries) / len(entries)
        r.worst_date = max(entries, key=lambda d: d.oob_min).date_key
        # The label the campaign earns most often; ties break toward the worst day.
        counts: dict[str, int] = {}
        for d in entries:
            counts[d.diagnosis] = counts.get(d.diagnosis, 0) + 1
        worst = max(entries, key=lambda d: d.oob_min).diagnosis
        r.dominant_diagnosis = max(counts, key=lambda k: (counts[k], k == worst))

        priced = [d.lost["lost_spend"] for d in entries
                  if d.lost and d.lost.get("lost_spend") is not None]
        r.total_lost_spend = sum(priced) if priced else None
        sales = [d.lost["lost_sales"] for d in entries
                 if d.lost and d.lost.get("lost_sales") is not None]
        r.total_lost_sales = sum(sales) if sales else None
        out.append(r)

    out.sort(key=lambda r: -r.chronic_score)
    return out

File: src/lib/aggregate.ts
/**
 * Roll campaign-days up across multiple exports to find chronic offenders.
 *
 * A campaign out of budget 8 hours a day for two weeks is a bigger problem than
 * one that spiked to 23 hours once, so `chronicScore` weights recurrence and
 * streak length alongside average severity.
 */

import { inHours, oobHours, pausedHours } from './scoring';
import type { CampaignDay } from './scoring';

const OOB_DAY_THRESHOLD_MIN = 60; // a day "counts" once an hour is lost
const STREAK_CEILING = 7;

const W_RECURRENCE = 0.4;
const W_MEAN = 0.35;
const W_STREAK = 0.25;

export interface DayPoint {
  dateKey: string;
  oobHours: number;
  inHours: number;
  pausedHours: number;
  episodes: number;
  severity: number;
  firstOobMin: number | null;
}

export interface CampaignRollup {
  campaign: string;
  daysObserved: number;
  daysWithOob: number;
  totalOobHours: number;
  meanOobHours: number;
  meanInHours: number; // hours per day the campaign could actually spend
  meanPausedHours: number;
  medianOobHours: number;
  maxOobHours: number;
  totalEpisodes: number;
  meanFirstOobMin: number | null;
  recurrenceRate: number;
  streakCurrent: number;
  streakMax: number;
  trendSlope: number;
  chronicScore: number;
  meanSeverity: number;
  dominantDiagnosis: string;
  worstDate: string;
  totalLostSpend: number | null;
  totalLostSales: number | null;
  perDay: DayPoint[];
}

export function trendLabel(r: CampaignRollup): 'flat' | 'worsening' | 'improving' {
  if (Math.abs(r.trendSlope) < 0.05) return 'flat';
  return r.trendSlope > 0 ? 'worsening' : 'improving';
}

function median(values: number[]): number {
  const s = [...values].sort((a, b) => a - b);
  const mid = Math.floor(s.length / 2);
  return s.length % 2 ? s[mid] : (s[mid - 1] + s[mid]) / 2;
}

/** Least-squares slope of y against its index. Zero for fewer than 2 points. */
function olsSlope(values: number[]): number {
  const n = values.length;
  if (n < 2) return 0;
  const meanX = (n - 1) / 2;
  const meanY = values.reduce((s, v) => s + v, 0) / n;
  let denom = 0;
  for (let i = 0; i < n; i++) denom += (i - meanX) ** 2;
  if (denom === 0) return 0;
  let numer = 0;
  for (let i = 0; i < n; i++) numer += (i - meanX) * (values[i] - meanY);
  return numer / denom;
}

/** First element with the largest key, matching Python's `max(..., key=...)`. */
function argmax<T>(items: T[], key: (item: T) => number): T {
  let best = items[0];
  let bestKey = key(best);
  for (const item of items.slice(1)) {
    const k = key(item);
    if (k > bestKey) {
      best = item;
      bestKey = k;
    }
  }
  return best;
}

export function rollup(days: CampaignDay[]): CampaignRollup[] {
  const byCampaign = new Map<string, CampaignDay[]>();
  for (const d of days) {
    const list = byCampaign.get(d.campaign);
    if (list) list.push(d);
    else byCampaign.set(d.campaign, [d]);
  }

  const out: CampaignRollup[] = [];
  for (const [campaign, entries] of byCampaign) {
    entries.sort((a, b) => (a.dateKey < b.dateKey ? -1 : a.dateKey > b.dateKey ? 1 : 0));
    const hours = entries.map(oobHours);
    const firsts = entries
      .map((d) => d.firstOobMin)
      .filter((v): v is number => v !== null);

    const perDay: DayPoint[] = entries.map((d) => ({
      dateKey: d.dateKey,
      oobHours: oobHours(d),
      inHours: inHours(d),
      pausedHours: pausedHours(d),
      episodes: d.episodesMerged,
      severity: d.severity,
      firstOobMin: d.firstOobMin,
    }));

    const daysWithOob = entries.filter((d) => d.oobMin >= OOB_DAY_THRESHOLD_MIN).length;
    const totalOobHours = hours.reduce((s, v) => s + v, 0);

    let streak = 0;
    let streakMax = 0;
    for (const d of entries) {
      if (d.oobMin >= OOB_DAY_THRESHOLD_MIN) {
        streak += 1;
        streakMax = Math.max(streakMax, streak);
      } else {
        streak = 0;
      }
    }

    const worst = argmax(entries, (d) => d.oobMin);
    // The label the campaign earns most often; ties break toward the worst day.
    const counts = new Map<string, number>();
    for (const d of entries) counts.set(d.diagnosis, (counts.get(d.diagnosis) ?? 0) + 1);
    let dominant = '';
    let dominantKey: [number, number] = [-1, -1];
    for (const [label, n] of counts) {
      const k: [number, number] = [n, label === worst.diagnosis ? 1 : 0];
      if (k[0] > dominantKey[0] || (k[0] === dominantKey[0] && k[1] > dominantKey[1])) {
        dominant = label;
        dominantKey = k;
      }
    }

    const pricedSpend = entries
      .map((d) => d.lost?.lostSpend)
      .filter((v): v is number => v !== null && v !== undefined);
    const pricedSales = entries
      .map((d) => d.lost?.lostSales)
      .filter((v): v is number => v !== null && v !== undefined);

    const meanOobHours = totalOobHours / entries.length;
    const chronicScore =
      100 *
      (W_RECURRENCE * (daysWithOob / entries.length) +
        W_MEAN * Math.min(meanOobHours / 24, 1) +
        (W_STREAK * Math.min(streakMax, STREAK_CEILING)) / STREAK_CEILING);

    out.push({
      campaign,
      daysObserved: entries.length,
      daysWithOob,
      totalOobHours,
      meanOobHours,
      meanInHours: entries.reduce((s, d) => s + inHours(d), 0) / entries.length,
      meanPausedHours: entries.reduce((s, d) => s + pausedHours(d), 0) / entries.length,
      medianOobHours: median(hours),
      maxOobHours: Math.max(...hours),
      totalEpisodes: entries.reduce((s, d) => s + d.episodesMerged, 0),
      meanFirstOobMin: firsts.length
        ? firsts.reduce((s, v) => s + v, 0) / firsts.length
        : null,
      recurrenceRate: daysWithOob / entries.length,
      streakCurrent: streak,
      streakMax,
      trendSlope: olsSlope(hours),
      chronicScore,
      meanSeverity: entries.reduce((s, d) => s + d.severity, 0) / entries.length,
      dominantDiagnosis: dominant,
      worstDate: worst.dateKey,
      totalLostSpend: pricedSpend.length ? pricedSpend.reduce((s, v) => s + v, 0) : null,
      totalLostSales: pricedSales.length ? pricedSales.reduce((s, v) => s + v, 0) : null,
      perDay,
    });
  }

  out.sort((a, b) => b.chronicScore - a.chronicScore);
  return out;
}
